docs: update README with clearer explanations (#356)
threeport/threeport: docs/docs/concepts/application-orchestration.md

# Application Orchestration

> Application orchestration is an approach to software delivery that favors software engineering over configuration management.  We employ distributed software systems that incorporate request-driven architecture, event-driven architecture and service-oriented systems to manage the deployment and ongoing management of applications.

Reviewer: ```suggestion
Application orchestration is an approach to software delivery that favors software engineering over configuration management.  Similar to how user-facing applications are often designed, Threeport employs event-driven and service-oriented architecture to manage the deployment and ongoing management of applications.
```

I think `request-driven` is a subset of `event-driven`, so I consolidated on `event-driven`. Also made it a little more concise, dropped `distributed software system` since I think we can draw parallels to software engineering without it. We also mention the distributed nature already in `docs/docs/architecture/threeport-core.md`

Author: I don't see request-driven as a subset of event-driven.  In my mind they're distinctly different.  [This article](https://www.geeksforgeeks.org/system-design/request-driven-vs-event-driven-microservices/) mirrors my understanding.  That said, request-driven systems are mostly trivial and non-differentiating.  So I like the suggested change here.

In order to bridge this gap we have used several different workarounds:

1. Domain specific languages: HashiCorp Config Language (HCL) is a good example of this which introduces some control flow constructs to help manage complex configs.
2. Templating systems: Helm is an example of a system that uses a templating language to produce large static configs.  Again control flow constructs are used to produce conditional and looping logic to manage vast configurations.
3. Overlay systems: Kustomize is an example of an overlay system that takes a base configuration, then adds, removes or replace elements for different use cases.

The problem with both DSLs, templates and overlays is that they manage logical constructs poorly when compared to general purpose programming languages.

The end result of config management systems is a rapid deterioration of maintainability as complexity grows.  Because of their limitations, they become littered with workarounds, complex combinations and enormous text file configs that are very challenging for humans to manage.

## Application Orchestration Approach

> With Threeport, we use a software engineering approach to solving the challenge of software delivery.  Request-driven, event-driven and service oriented architectures are powerful and effective when executed well.  We are applying these principles to application platforms.

Reviewer: ```suggestion
With Threeport, we use a software engineering approach to solving the challenge of software delivery.  Event-driven and service-oriented architectures are powerful and effective when executed well.  We are applying these principles to application platforms.
```

One way to think about Threeport is that it is Ruby on Rails for application platforms.

The table below helps illustrate the difference between the approaches.

|  | **Config Management** | **Application Orchestration** |
|---|---|---|
| Team Sepciality | DevOps | Platform engineering |

> | Primary Focus | Text configs, templates, pipelines | Software controllers, Threeport modules |

Reviewer: ```suggestion
| Primary Focus | Text configs, templates, pipelines | Threeport modules (software service or controller) |
```
